### PR TITLE
Jetpack: Use toggle for stats module

### LINKED ISSUE #22216: Jetpack Settings: Stats notice should be converted to a module toggle
Right now, if a user has their Stats Jetpack module deactivated, we display a notice that contains a button to activate that module:

This is inconsistent with the rest of the settings for modules where we display a module toggle, for example **Ads**:

In order to make it less scary and more consistent, we should turn the notice into a card with a module activation toggle.

cc @rickybanister

### PR DESCRIPTION
**Pull request by sirreal**
Change the Jetpack `stats` module control from a single one-way button to a toggle.

#### Disabled

before-dis

#### Enabled

before-en

#### Disabled

after-dis

#### Enabled

after-en

## Testing
1. Set up a Jetpack site
1. Visit https://calypso.live/settings/traffic?branch=fix/22216-jp-stats-toggle
1. Verify that the new toggle behaves as expected, settings are correctly modified and persisted.

Fixes #22216

### CODE REVIEW

File: client/my-sites/site-settings/jetpack-site-stats.jsx [hunk 1]
```diff
@@ -115,34 +113,34 @@ class JetpackSiteStats extends Component {
...
-			<Banner
-				title={ translate( 'Site Stats module is disabled' ) }
-				description={ translate(
-					'Enable this to see detailed information about your traffic, likes, comments, and subscribers.'
+		const header = (
+			<JetpackModuleToggle
+				siteId={ siteId }
+				moduleSlug="stats"
+				label={ translate(
+					'See detailed information about your traffic, likes, comments, and subscribers.'
```
Comment (tyxla) [inline client/my-sites/site-settings/jetpack-site-stats.jsx]: I would really love it if we could keep this on a single line. Maybe `See details` instead of `See detailed information`, or `Monitor` instead of `See detailed information about`?

File: client/my-sites/site-settings/jetpack-site-stats.jsx [hunk 2]
```diff
@@ -115,34 +113,34 @@ class JetpackSiteStats extends Component {
...
 				) }
-				callToAction={ translate( 'Enable' ) }
-				onClick={ this.handleStatsActivationButton }
-				event={ 'site_stats_module_enable_banner' }
-				icon="stats"
+				disabled={ activatingStatsModule || moduleUnavailable }
```
Comment (tyxla) [inline client/my-sites/site-settings/jetpack-site-stats.jsx]: `JetpackModuleToggle` doesn't need the `activatingStatsModule` in its `disabled` - it will check for that anyway. This will allow us to also remove the `activateModule` action creator altogether.

File: client/my-sites/site-settings/jetpack-site-stats.jsx [hunk 3]
```diff
@@ -115,34 +113,34 @@ class JetpackSiteStats extends Component {
...
-					'Enable this to see detailed information about your traffic, likes, comments, and subscribers.'
+		const header = (
+			<JetpackModuleToggle
+				siteId={ siteId }
+				moduleSlug="stats"
+				label={ translate(
+					'See detailed information about your traffic, likes, comments, and subscribers.'
 				) }
-				callToAction={ translate( 'Enable' ) }
-				onClick={ this.handleStatsActivationButton }
```
Comment (tyxla) [inline client/my-sites/site-settings/jetpack-site-stats.jsx]: With the removal of this call, the `handleStatsActivationButton` method is now unnecessary.

File: client/my-sites/site-settings/jetpack-site-stats.jsx [hunk 4]
```diff
@@ -4,31 +4,29 @@
...
-import { getStatsPathForTab } from 'lib/route';
+import SectionHeader from 'components/section-header';
+import { activateModule } from 'state/jetpack/modules/actions';
 import { getSelectedSiteId, getSelectedSiteSlug } from 'state/ui/selectors';
 import { getSiteRoles } from 'state/site-roles/selectors';
-import { activateModule } from 'state/jetpack/modules/actions';
+import { getStatsPathForTab } from 'lib/route';
 import {
 	isActivatingJetpackModule,
 	isJetpackModuleActive,
```
Comment (tyxla) [inline client/my-sites/site-settings/jetpack-site-stats.jsx]: Do we want to hide the `View your site stats` link if the module is inactive or unavailable? cc @Automattic/jetpack-design

**Comment by rickybanister** [comment]
Stats will still be toggled to the 'on' position by default on install, correct (even without Jumpstarting...)?

We don't really want people to turn them off unless they have a GDPR concern or something. 

Are there any implications with Atomic sites or other circumstances where we don't want to allow a user to modify the setting? I can't remember if @oskosk or @ebinnion was working on that recently. 

Design-wise it looks great, aligning it with the way spell check works makes perfect sense. Thanks all.

**Comment by ebinnion** [comment]
I don't imagine there being an issue. The UI work on Calypso hasn't been started for forcing a module on/off. We probably want to copy the toggle UI for stats to Jetpack though.

File: client/my-sites/site-settings/jetpack-site-stats.jsx (hunk 4 above)
Comment (keoshi) [reply client/my-sites/site-settings/jetpack-site-stats.jsx]: @tyxla I'd keep the link — it's still useful for discovery and if there were any stats collected before it's a major reason to re-enable stats. In case it's inactive, we could add a notice near the stats itself letting the user know.

What do you mean by “unavailable” though? Are there certain scenarios where I can't turn on stats?

Comment (rickybanister) [reply client/my-sites/site-settings/jetpack-site-stats.jsx]: Modules can be disabled in code by a developer, or by a host. We are working to make sure our UI responds to those conditions.

Comment (keoshi) [reply client/my-sites/site-settings/jetpack-site-stats.jsx]: I'd still keep it unless we can tell:

- We haven't collected any stats before for this site.
- We can tell the stats module is “unavailable” rather than anything else rendering it “inactive”.

**Comment by MichaelArestad** [comment]
I don't think we should make this change. We don't want to encourage folks to disable certain modules. Making it easy to disable will do just that.

**Comment by tyxla** [comment]
I thought it was decided that for now we're going with module toggles instead of activation notices - see p1518015413000456-slack-jetpack-design and the thread in p8oabR-9V-p2 #comment-1311 for design discussions with @rickybanister that confirmed this. IMHO that approach would actually make it way more consistent with the rest of modules that are handled throughout Jetpack settings in Calypso.

We could work on better implicit activation (automatically enabling a module once the user wants to use it in any way), but this could have many different cases and IMHO could be an entire separate project to make it happen for all modules that make sense.

File: client/my-sites/site-settings/jetpack-site-stats.jsx (hunk 3 above)
Comment (sirreal) [reply client/my-sites/site-settings/jetpack-site-stats.jsx]: 👍 nice catch

File: client/my-sites/site-settings/jetpack-site-stats.jsx (hunk 2 above)
Comment (sirreal) [reply client/my-sites/site-settings/jetpack-site-stats.jsx]: 🎉 d520fceb37d6b050baf54f0994d0e831c9f9b2f0

**Comment by sirreal** [comment]
I've applied the suggestions from @tyxla, significant code reduction (💯 🙇)

**Comment by designsimply** [comment]
Tested and confirmed I can see a toggle option for "Site stats" in Calypso > Settings > Traffic for a Jetpack site and that saving changes to that option persist after refreshing the page.

I noticed that when I turn off the "Site stats" setting in Calypso, I can still see Stats in the left menu on Calypso but I cannot see Site Stats in the left menu in WP Admin. Is that intentional? (It does make sense to me that Stats should always be accessible in Calypso because that is the main route I know of to get to Insights and Activity.)

screen shot 2018-03-07 at wed mar 7 4 25 12 pm
<sup>Seen at https://calypso.live/settings/traffic/madefortesting.com running WordPress 4.9.4 and Jetpack 5.9 using Firefox 58.0.2 on macOS 10.13.3.</sup>

screen shot 2018-03-07 at wed mar 7 4 25 16 pm
<sup>Seen at https://madefortesting.com/wp-admin/admin.php?page=jetpack#/traffic running WordPress 4.9.4 and Jetpack 5.9 using Firefox 58.0.2 on macOS 10.13.3.</sup>

**Comment by designsimply** [comment]
(Out of curiosity) is it possible to disable "Site stats" from WP Admin? I looked but couldn't find a way to disable it from WP Admin > Jetpack > Settings > Traffic.

**Comment by tyxla** [comment]
@MichaelArestad has some updated feedback for all the toggle cases - see https://github.com/Automattic/wp-calypso/pull/23034#issuecomment-371864065

TL;DR: we have a green light to go with them, as long as there aren't any additional concerns.

**Comment by sirreal** [comment]
> I noticed that when I turn off the "Site stats" setting in Calypso, I can still see Stats in the left menu on Calypso but I cannot see Site Stats in the left menu in WP Admin. Is that intentional?

@designsimply Thanks for the feedback. I believe that "Stats" does need to remain in the menu because it provides access to more functionality that just the site stats provided by this module.

It might be worth opening a separate issue to provide some context and links to enable the module from the stats page if the stats module is deactivated.
